diff --git a/dist/worker.cjs b/dist/worker.cjs
index f1ddf70b4ce08b1592ee6f40d77259d084263ccc..412d2fd0d6f9302312c45989f70fdc9705c24883 100644
--- a/dist/worker.cjs
+++ b/dist/worker.cjs
@@ -576,6 +576,7 @@ class AgentServer extends import_events.EventEmitter {
               canPublish: this.#opts.permissions.canPublish,
               canSubscribe: this.#opts.permissions.canSubscribe,
               canPublishData: this.#opts.permissions.canPublishData,
+              canPublishSources: this.#opts.permissions.canPublishSources,
               canUpdateMetadata: this.#opts.permissions.canUpdateMetadata,
               hidden: this.#opts.permissions.hidden,
               agent: true
diff --git a/dist/worker.js b/dist/worker.js
index 8b91c7fa57c3ae4e167023fd1b3ea0c9b120a6a0..a80dd63cef12a05956931e30ec45293c1c9047fe 100644
--- a/dist/worker.js
+++ b/dist/worker.js
@@ -550,6 +550,7 @@ class AgentServer extends EventEmitter {
               canPublish: this.#opts.permissions.canPublish,
               canSubscribe: this.#opts.permissions.canSubscribe,
               canPublishData: this.#opts.permissions.canPublishData,
+              canPublishSources: this.#opts.permissions.canPublishSources,
               canUpdateMetadata: this.#opts.permissions.canUpdateMetadata,
               hidden: this.#opts.permissions.hidden,
               agent: true
