// convex/dist/cjs/server/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var server_exports = {};
__export(server_exports, {
  HttpRouter: () => import_router.HttpRouter,
  ROUTABLE_HTTP_METHODS: () => import_router.ROUTABLE_HTTP_METHODS,
  actionGeneric: () => import_registration_impl.actionGeneric,
  anyApi: () => import_api.anyApi,
  componentsGeneric: () => import_components.componentsGeneric,
  createFunctionHandle: () => import_components.createFunctionHandle,
  cronJobs: () => import_cron.cronJobs,
  currentSystemUdfInComponent: () => import_components2.currentSystemUdfInComponent,
  defineApp: () => import_components.defineApp,
  defineComponent: () => import_components.defineComponent,
  defineSchema: () => import_schema.defineSchema,
  defineTable: () => import_schema.defineTable,
  filterApi: () => import_api.filterApi,
  getFunctionAddress: () => import_components3.getFunctionAddress,
  getFunctionName: () => import_api.getFunctionName,
  httpActionGeneric: () => import_registration_impl.httpActionGeneric,
  httpRouter: () => import_router.httpRouter,
  internalActionGeneric: () => import_registration_impl.internalActionGeneric,
  internalMutationGeneric: () => import_registration_impl.internalMutationGeneric,
  internalQueryGeneric: () => import_registration_impl.internalQueryGeneric,
  makeFunctionReference: () => import_api.makeFunctionReference,
  mutationGeneric: () => import_registration_impl.mutationGeneric,
  queryGeneric: () => import_registration_impl.queryGeneric
});
module.exports = __toCommonJS(server_exports);
__reExport(server_exports, require("./database.js"), module.exports);
var import_registration_impl = require("./impl/registration_impl.js");
__reExport(server_exports, require("./pagination.js"), module.exports);
__reExport(server_exports, require("./search_filter_builder.js"), module.exports);
__reExport(server_exports, require("./storage.js"), module.exports);
var import_cron = require("./cron.js");
var import_router = require("./router.js");
var import_api = require("./api.js");
var import_components = require("./components/index.js");
var import_components2 = require("./components/index.js");
var import_components3 = require("./components/index.js");
var import_schema = require("./schema.js");
